controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,241 @@
+        this.gcName = gcName;
+        this.leaderLatch = new AtomicReference<>();
+        this.leaderPath = String.format(LEADER_PATH, gcName);
+        this.watch = new AtomicReference<>();
+        this.batchPath = String.format(BATCH_PATH, gcName);
+        this.zkStoreHelper = zkStoreHelper;
+        this.gcProcessing = gcProcessing;
+        this.periodInMillis = gcPeriod.toMillis();
+        this.gcExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.gcLoop = new AtomicReference<>();
+    }
+
+    @Override
+    protected void doStart() {
+        long quantized = computeNewBatch();
+        byte[] b = getBytes(quantized);
+        // 1. create znode for storing latest batch id. If the batch id exists, get the value from the store.
+        // We will later register watch on the path and keep receiving any changes to its value.
+        zkStoreHelper.createZNodeIfNotExist(batchPath, b)
+                .thenCompose(x -> zkStoreHelper.getData(batchPath))
+                .thenAccept(data -> {
+                    currentBatch.set(BitConverter.readLong(data.getData(), 0));
+                })
+                .thenRun(this::initialize)
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                    latch.complete(null);
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        latch.thenAccept(v -> {
+            if (gcLoop.get() != null) {
+                gcLoop.get().cancel(true);
+                gcLoop.get().whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStopped();
+                    }
+                });
+            } else {
+                notifyStopped();
+            }
+        });
+    }
+
+    public Long getLatestBatch() {
+        return currentBatch.get();
+    }
+
+    @SneakyThrows(Exception.class)
+    private void initialize() {
+        // 1. register watch on batch path
+        watch.set(registerWatch(batchPath));
+
+        // 2. attempt to acquire leadership
+        leaderLatch.set(electLeader(leaderPath, acquiredLeadership));
+
+        // 3. if this acquires leadership, then schedule periodic garbage collection.
+        // Note: This will attempt first GC only after one period elapses. This will ensure that even if there were 
+        // back to back leadership changes with each having their own clock skews, a completed transaction record
+        // will at least be present for one GC period. 
+        // The downside is, if there are controller failures before GC period, then GC keeps getting delayed perpetually. 
+        gcLoop.set(acquiredLeadership.thenCompose(x -> Futures.loop(this::isRunning, () -> Futures.delayedFuture(this::process,
```

Comment:
acquiredLeadership is available to the outside to technically an outside class could fail it and mess with the gcLoop. It would be better if the object that was available to the outside were a separate future that was just released by this one.